### ISSUE TITLE
Format document doesn't indent the first statements after if correctly

### ISSUE THREAD

**Issue by VolkmarR** [X-Sharp/XSharpPublic#552]
**To Reproduce**
1. Extract the file [If_bugs.zip](https://github.com/X-Sharp/XSharpPublic/files/5771341/If_bugs.zip) and open the file If_bugs.prg in Visual Studio
2. Execute Format Document (Ctrl + k + d)
3. Now the statement in line 10 is incorrectly indented
4. Remove the line 11
5. Execute Format Document (Ctrl + k + d)
6. Now the statement in line 10 is correctly indented

**Expected behavior**
Multiple executions of Format Document should always produce the same formatting

**Desktop (please complete the following information):**
- OS: Win10
- Visual Studio Version: 2019 (16.8.3)
- X#: 2.7

**Comment by VolkmarR** [comment]
This is the content of the file in the zip:

```
class clsPasswordGenerator2 

    method __GeneratePasswordSimple() as string strict
        
        local cPassword := ""
        local nLastPart := 1
        
        // the indentation of the statement >cPassword += "a"< is constantly incorrect, when you execute format documentent 
        if nLastPart == 1
            cPassword += "a"
            // If you remove this comment, then the cPassword assignment will be indented correctly
        elseif nLastPart == 2
            cPassword += "b"
        endif
        
        return cPassword
        
        
        // the indentation of "end class" changes sometimes, when you execute format documentent multiple times
    END CLASS
```

**Comment by cpyrgas** [comment]
The main problem is fixed, but I think I know what's causing the reported problem with END CLASS:
1.Execute format document, now END CLASS is moved at column 1 (correct)
2.Go at the end of the END CLASS line and press <Enter>, now END CLASS is moved a tab to the right, aligning with "METHOD"
3.Repeat 1-2, this always happens.
4.Add a "END METHOD" at the end of the method code. From now on, END CLASS is already being correctly aligned to the beginning of the line

**Comment by fforay** [comment]
This one is solved with new Formatting process

**Comment by cpyrgas** [comment]
The problem with format document is fixed, but the problem with putting back END CLASS to the wrong place is still happening.

**Comment by fforay** [comment]
Are you sure ? With the Latest version, this works fine.
By latest version, I mean the one in the feature/Intellisense branch, not the one in the latest release (wich is the old one ;) )

**Comment by VolkmarR** [comment]
Let me know, if you need an additional tester for the formatting.

**Comment by cpyrgas** [comment]
Fabrice, I am using the latest commit which was done around 2-3 hours before now. Could it be you have uncommitted changes?

Just to be totally clear:
- Format document works nicely!
- When I go to the END CLASS and hit <enter>, then END CLASS in indented to the right

I tried a few different formatting options, but I always get this behavior

Volkmar, as soon as a couple more issues are taken care of, I will send you a net set of X# dll files to install in VS so you can also test the changes!

**Comment by fforay** [comment]
Chris, I cannot reproduce it, but....
Formatting the document with Ctrl-K + Ctrl-D is a different process than "Auto" formatting/indenting when typing.
The "new" process concern the "Format Document", which is ok, right ?
So... I will investigate the indenting when typing.

**Comment by cpyrgas** [comment]
Fabrice, yes, the format document command itself works very nicely, at least with the code snippet of this ticket!

The problem I see is what I described above, but I see Git has modified my message, let me try again:

"When I go to the END CLASS and hit the ENTER key, then END CLASS is indented one tab to the right"

But I am seeing more issues like that with indenting, that I know you and Robert cannot reproduce, and I am not sure what is causing the different behavior in my machine (and I am not sure if it happens in other people's machines). So if you do not manage to reproduce it after some time, please forget about it, I will send an update to Volkmar to test your fixes and if he considers the problem fixed, we can just close the ticket.

**Comment by RobertvanderHulst** [comment]
The indenting code seems to match the END of END CLASS with the METHOD keyword.
I'll change that so that when end is followed by another keyword (END CLASS, END INTERFACE, END LOCK) then it will match

**Comment by cpyrgas** [comment]
Indeed, it's a lot better now, but there are still cases when END CLASS is being wrongly relocated. I tried this, and when I press enter while in the last line, again END CLASS is being indented one tab to the right, I think it's confusing the nested class with the actual main class that it belongs to:

```
CLASS clsPasswordGenerator2 

	CLASS InnerClass
	END CLASS

	METHOD __GeneratePasswordSimple() AS STRING STRICT
        
END CLASS

```

**Comment by cpyrgas** [comment]
I'm still seeing the problem described above https://github.com/X-Sharp/XSharpPublic/issues/552#issuecomment-1019880876

**Comment by cpyrgas** [comment]
This looks fully fixed now.
